### PR TITLE
BUGFIX: Declare properties in proxy classes to prevent PHP 8.2 deprecations

### LINKED ISSUE #2946: BUG: (PHP 8.2 Support) Creation of dynamic property [...]::$Flow_Injected_Properties is deprecated
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Running a basic Neos.Demo installation on PHP 8.2 currently leads to this (repeated) PHP deprecation message:
```
Deprecated: Creation of dynamic property Neos\Flow\ResourceManagement\Streams\ResourceStreamWrapper::$Flow_Injected_Properties is deprecated in /tmp/neos/Development/SubContextddev/Cache/Code/Flow_Object_Classes/Neos_Flow_ResourceManagement_Streams_ResourceStreamWrapper.php on line 591
```

Indeed, dynamic properties have been deprecated in PHP 8.2:
>  The creation of dynamic properties is deprecated to help avoid mistakes and typos, unless the class opts in by using the `#[\AllowDynamicProperties]` attribute. `stdClass` allows dynamic properties.
>
> Usage of the __get/__set magic methods is not affected by this change.

(see: https://www.php.net/releases/8.2/en.php)

### Expected Behavior

There is no deprecation message.

### Steps To Reproduce

* Set up a Neos Distribution running Neos.Demo on PHP 8.2
* See deprecation message on homepage

### Environment

```markdown
- Flow: 8.2
- PHP: 8.2
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by paxuclus**
This patch resolves deprecation notices on PHP 8.2:

## Dynamic properties in proxy classes
There are some properties that are dynamically declared in proxy classes:
* `Flow_Injected_Properties`
* `Flow_Object_PropertiesToSerialize`
* `Flow_Persistence_RelatedEntities`: this is mostly used inside of the `ObjectSerializationTrait`, so I thought it might make more sense to declare it there instead of adding a property using the `ProxyClassBuilder`

Resolves #2946

## `parent` inside of closures
There are some methods that will be checked against `is_callable` with `parent::`:
* `parent::Flow_Aop_Proxy_buildMethodsAndAdvicesArray`
* `parent::__wakeup`

**Review instructions**
* Set up a flow distribution on PHP 8.2
* Run tests and make sure no deprecation warnings are thrown

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked wit `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by kitsunet** [comment]
Good idea +++ But I guess we need to ensure those properties do not get serialised seeing the test issues.

**Comment by paxuclus** [comment]
@kitsunet the tests pass now. What about the target branch? Is 8.0 fine or should this go into 7.3?

**Comment by markusguenther** [comment]
Good to see that coming. Had today the issue that my Neos 9.0 tryout throw the error:

```
PHP Deprecated:  Creation of dynamic property Neos\Flow\ResourceManagement\Streams\ResourceStreamWrapper::$Flow_Injected_Properties is deprecated in /Volumes/Development/Playground/neos-90/Data/Temporary/Development/Cache/Code/Flow_Object_Classes/Neos_Flow_ResourceManagement_Streams_ResourceStreamWrapper.php on line 591
```

Fixed that now temporary with `#[\AllowDynamicProperties]` in the class, but that is just my workaround :D

File: Neos.Flow/Classes/ObjectManagement/Proxy/ObjectSerializationTrait.php [hunk 1]
```diff
@@ -116,7 +123,7 @@ private function Flow_searchForEntitiesAndStoreIdentifierArray(string $path, mix
                 $this->Flow_searchForEntitiesAndStoreIdentifierArray($path . '.' . $key, $value, $originalPropertyName);
             }
         } elseif ($propertyValue instanceof PersistenceMagicInterface && !Bootstrap::$staticObjectManager->get(PersistenceManagerInterface::class)->isNewObject($propertyValue) || $propertyValue instanceof DoctrineProxy) {
-            if (!property_exists($this, 'Flow_Persistence_RelatedEntities') || !is_array($this->Flow_Persistence_RelatedEntities)) {
+            if (!property_exists($this, 'Flow_Persistence_RelatedEntities') || !isset($this->Flow_Persistence_RelatedEntities) || !is_array($this->Flow_Persistence_RelatedEntities)) {
```
Comment (mhsdesign) [inline Neos.Flow/Classes/ObjectManagement/Proxy/ObjectSerializationTrait.php]: It seems we don’t need `property_exists` anymore ;)

and we should def look up if property_exists is used elsewhere ein the codebase (as it might not work anymore)

**Comment by mhsdesign** [comment]
After discussing I think we should target 7.3 https://neos-project.slack.com/archives/C04Q0TC15/p1677852202036989

File: Neos.Flow/Classes/ObjectManagement/DependencyInjection/PropertyInjectionTrait.php [hunk 2]
```diff
@@ -16,6 +16,8 @@
  */
 trait PropertyInjectionTrait
 {
+    protected $Flow_Injected_Properties = [];
```
Comment (bwaidelich) [inline Neos.Flow/Classes/ObjectManagement/DependencyInjection/PropertyInjectionTrait.php]: I think this can have a type hint!?
```suggestion
    protected array $Flow_Injected_Properties = [];
```

File: Neos.Flow/Classes/ObjectManagement/Proxy/ObjectSerializationTrait.php [hunk 3]
```diff
@@ -28,6 +28,10 @@
  */
 trait ObjectSerializationTrait
 {
+    protected $Flow_Object_PropertiesToSerialize = [];
```
Comment (bwaidelich) [inline Neos.Flow/Classes/ObjectManagement/Proxy/ObjectSerializationTrait.php]:
```suggestion
    protected array $Flow_Object_PropertiesToSerialize = [];
```

File: Neos.Flow/Classes/ObjectManagement/Proxy/ObjectSerializationTrait.php [hunk 4]
```diff
@@ -28,6 +28,10 @@
...
+
+    protected $Flow_Persistence_RelatedEntities = null;
```
Comment (bwaidelich) [inline Neos.Flow/Classes/ObjectManagement/Proxy/ObjectSerializationTrait.php]:
```suggestion
    protected ?array $Flow_Persistence_RelatedEntities = null;
```

File: Neos.Flow/Tests/Functional/ObjectManagement/ProxyCompilerTest.php [hunk 5]
```diff
@@ -194,7 +194,9 @@ public function complexPropertyTypesArePreserved()
         $reflectionClass = new ClassReflection(Fixtures\PHP8\ClassWithUnionTypes::class);
         /** @var PropertyReflection $property */
         foreach ($reflectionClass->getProperties() as $property) {
-            if ($property->getName() !== 'propertyA' && $property->getName() !== 'propertyB') {
+            if (str_starts_with($property->getName(), 'property') &&
```
Comment (bwaidelich) [inline Neos.Flow/Tests/Functional/ObjectManagement/ProxyCompilerTest.php]: Why is this addition required?

Comment (paxuclus) [reply Neos.Flow/Tests/Functional/ObjectManagement/ProxyCompilerTest.php]: By explicitly declaring the `Flow_*` properties in the traits, `$reflectionClass->getProperties()` will now also return those new properties, which breaks this test. 
I'm not really happy with this change but simply checking for the property name was the easiest way to fix the test.

This test could also be refactored to be a bit stricter with the property names that are checked (e.g. only look at `$propertyC`, `$propertyD` and `$propertyE` of the [fixture class](https://github.com/neos/flow-development-collection/blob/8.3/Neos.Flow/Tests/Functional/ObjectManagement/Fixtures/PHP8/ClassWithUnionTypes.php#L28))

Comment (kitsunet) [reply Neos.Flow/Tests/Functional/ObjectManagement/ProxyCompilerTest.php]: Incidentally I just created the same fix and stumbled on the same test failure, it makes sense.

**Comment by kitsunet** [comment]
Hey, we just discussed this further on our code sprint and I would move this up to the 8.0 branch so we can be sure to have PHP 8.x requirements in place for this and add some small code cleanups I did along the way.
